File: XCPC/Others/Baozii_Cup2/G.py
import sys

for _ in range(int(input())):
    cnt = [0] * 64

    for i in range(1, 64):
        print("?", pow(2,i))
        sys.stdout.flush()
        cnt[i] = int(input())

    mx = 0
    mn = 64

    for i in range(1, 64):
        mx = max(mx, cnt[i])
        mn = min(mn, cnt[i])

    ans = 0

    if(mx == mn):
        if mx == 1:
            ans = 0
        elif mx == 2:
            ans = 1
        else:
            ans = pow(2, 64) - 1
    
    else:
        tnc = 0
        for i in range(1, 64):
            if cnt[i] == mn:
                tnc += 1
                ans += pow(2, i)
        
        if tnc != mn + 1:
            ans += 1
    
    print("!", ans)
    sys.stdout.flush()

File: XCPC/Others/Baozii_Cup2/G.cpp
#include<bits/stdc++.h>
using namespace std;
using pii = pair<int, int>;
using ll = long long;
using ull = unsigned long long;

int ask(ull x)
{
    cout << "? " << x << endl;
    cout.flush();
    int res;
    cin >> res;
    return res;
}

void solve()
{
    ull ans = 0;
    vector<int> cnt(64);
    for(int i = 63; i >= 1; i--){
        cnt[i] = ask(1ull << i);
    }
    int mx = *max_element(cnt.begin() + 1, cnt.end());
    int mn = *min_element(cnt.begin() + 1, cnt.end());
    
    if(mx == mn){
        if(mx == 1) ans = 0;
        else if(mx == 2) ans = 1; 
        else ans = ULLONG_MAX; //... 糖丸了 幸好拿py写了一遍
    }else{
        int tnc = 0;
        for(int i = 63; i >= 1; i--){
            if(cnt[i] == mn){
                tnc++;
                ans |= 1ull << i;
            }
        }
        if(tnc != mn + 1) ans |= 1, tnc++;
        assert(tnc == mx - 1);
    }
    cout << "! " << ans << endl;
    cout.flush();
}
int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr), cout.tie(nullptr);
    int T = 1;
    cin >> T;
    while(T--){
        solve();
    }
    return 0;
}